fix confusing code for http.Header handling

aead opened the pull request:
The extractMetadataFromHeader function should return an error if
the http.Header contains a non-canonicalized key.

## Motivation and Context
The problem is that if someone sets a key of http.Header manually (through a map access)
like `request["a-lower-case-key"] = "something"` the http.Header.Get function still searches for a canonicalized key. This can lead to hard-to-find bugs.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
fix header-to-metadat extraction

The extractMetadataFromHeader function should return an error if
the http.Header contains a non-canonicalized key. The reason is that
the keys can be manually set (through a map access) which can lead to
ugly bugs.

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -116,11 +116,10 @@ func extractMetadataFromHeader(header http.Header) map[string]string {
-		if strings.HasPrefix(cKey, "X-Amz-Meta-") {
-			metadata[cKey] = header.Get(key)
+		if strings.HasPrefix(key, "X-Amz-Meta-") {
+			metadata[key] = header.Get(key)
```
are you sure that key is always Canonical? @aead  - this code was added on purpose. and why is it confusing?

harshavardhana replied:
https://play.golang.org/p/hCJ6n1zLZQ 

```go
package main

import (
	"fmt"
	"net/http"
)

func main() {
	headers := http.Header{}
	headers.Set("content-type", "test")
	headers["x-amz-meta-x-amz-key"] = []string{"vvv"}
	for k, v := range headers {
		fmt.Println(k, v)
	}
}
```

How do you handle a situation like this? is it expected that `headers` is not modified by the caller?

harshavardhana replied:
So here is the reason why this is done.. https://play.golang.org/p/KRWnukcwVd

```go
package main

import (
	"fmt"
	"net/http"
	"strings"
)

func main() {
	headers := http.Header{}
	headers.Set("content-type", "test")
	headers["x-amz-meta-x-amz-key"] = []string{"vvv"}
	for k, v := range headers {
		key := http.CanonicalHeaderKey(k)
		if strings.HasPrefix(key, "X-Amz-Meta") {
			fmt.Println(key, v)
		}
	}
}
```

donatello commented on cmd/handler-utils.go:
```diff
@@ -116,16 +116,17 @@ func extractMetadataFromHeader(header http.Header) map[string]string {
-			metadata[cKey] = header.Get(key)
-		} else if strings.HasPrefix(key, "X-Minio-Meta-") {
-			metadata[cKey] = header.Get(key)
+		if key != http.CanonicalHeaderKey(key) {
```
I don't understand why this is needed - the only way to get a non-canonical header is if we add code that changes request headers. Request headers are not supposed to be modified - any code that does so is introducing bugs. So checking for this and returning errors to the user does not appear to be correct, as they could not have caused such a problem in the first place.

harshavardhana replied:
It is quite possible @donatello because this is a function and http.Header by nature is not protected. So anyone can modify and use it. This function is defining its minimum conditions not hoping that http.Header itself is never modified. 

IMO it is better to error out if we do indeed see an unexpected key since this function cannot operate in any other way.

deekoder replied:
Any thoughts @aead  further?

harshavardhana replied:
It's already addressed @deekoder look at the patch itself.

deekoder replied:
@y4m4 wanted to confirm with @aead. Please see Travis failed.

harshavardhana commented on cmd/handler-utils.go:
```diff
@@ -98,9 +98,9 @@ func path2BucketAndObject(path string) (bucket, object string) {
+func extractMetadataFromHeader(header http.Header) (map[string]string, error) {
 	if header == nil {
-		return nil
+		return nil, nil
```
Even this is `errInvalidArgument`

donatello commented on cmd/web-handlers.go:
```diff
@@ -497,7 +497,11 @@ func (web *webAPIHandlers) Upload(w http.ResponseWriter, r *http.Request) {
-	metadata := extractMetadataFromHeader(r.Header)
+	metadata, err := extractMetadataFromHeader(r.Header)
+	if err != nil {
+		writeErrorResponse(w, ErrInvalidMetadataDirective, r.URL)
```
The error being returned here is not correct - `extractMetadataFromHeader` returns an `ErrInvalidArgument` but we return an error about metadata directive? It should be an internal server error with a message about an invalid condition being found in the server. An invalid canonical header cannot be caused by the user or an application, so it is not correct so send such an error back to the user.

Commit:
fix header-to-metadata extraction

Return a InternalError if a non-canonicalized key is found in a http.Header.
Also log the error.
